jcl/SortKernels.mk

```diff
@@ -0,0 +1,68 @@
+###############################################################################
+# Copyright (c) 2018, 2018 IBM Corp. and others
+#
+# This program and the accompanying materials are made available under
+# the terms of the Eclipse Public License 2.0 which accompanies this
+# distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+# or the Apache License, Version 2.0 which accompanies this distribution and
+# is available at https://www.apache.org/licenses/LICENSE-2.0.
+#
+# This Source Code may also be made available under the following
+# Secondary Licenses when the conditions for such availability set
+# forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+# General Public License, version 2 with the GNU Classpath
+# Exception [1] and GNU General Public License, version 2 with the
+# OpenJDK Assembly Exception [2].
+#
+# [1] https://www.gnu.org/software/classpath/license.html
+# [2] http://openjdk.java.net/legal/assembly-exception.html
+#
+# SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+###############################################################################
+
+# Makefile to compile generated GPU code in SortKernels.cu.
+#
+# Example call:
+#   $(MAKE) -f SortKernels.mk GPU_BIN_DIR=$(GPU_BIN_DIR) GPU_TMP_DIR=$(GPU_TMP_DIR) NVCC=$(NVCC)
+
+# Verify required macros are defined.
+ifndef GPU_BIN_DIR
+$(error GPU_BIN_DIR is undefined)
+endif
+ifndef GPU_TMP_DIR
+$(error GPU_TMP_DIR is undefined)
+endif
+ifndef NVCC
+$(error NVCC is undefined)
+endif
+
+# Find all unique occurrences of compute_NN and sm_NN in the output
+# of 'nvcc --help'; we assume they are supported GPU architectures.
+GPU_ARCH_LIST := $(strip $(sort $(shell \
+	$(NVCC) --help \
+		| sed -e "s/[^a-z_0-9]/ /g" -e "y/ /\n/" \
+		| grep -E "^(compute|sm)_[0-9]+$$")))
+
+# Expands to a '--generate-code' option and value if both the arguments are
+# supported GPU architectures (possibly virtual) or nothing otherwise.
+# The next two macros are helpers: GPU_ARCH_TEST tests if the given architectures
+# are supported, GPU_ARCH_GENERATE is needed to hide the comma from the 'if' function.
+GPU_ARCH_CODE = \
+	$(if $(call GPU_ARCH_TEST,$1,$2), $(call GPU_ARCH_GENERATE,$1,$2))
+GPU_ARCH_TEST = \
+	$(and $(findstring $1,$(GPU_ARCH_LIST)), $(findstring $2,$(GPU_ARCH_LIST)))
+GPU_ARCH_GENERATE = \
+	--generate-code arch=$1,code=$2
+
+# Expands the the subset of '--generate-code' options supported by the version
+# of nvcc being used.
+GPU_ARCH_CODES := \
+	$(call GPU_ARCH_CODE,compute_20,compute_20) \
+	$(call GPU_ARCH_CODE,compute_20,sm_20) \
+	$(call GPU_ARCH_CODE,compute_30,compute_30) \
+	$(call GPU_ARCH_CODE,compute_30,sm_30) \
+	$(call GPU_ARCH_CODE,compute_35,sm_35)
```

Comment:
In case the -help output changes or isn't the same across all platforms, can you check if there are any matches for at least one of the compute values and cause a build error if there are no matches found?

Reply:
Probably the simplest way to do that would be to make the critical combination unconditional; that is we could replace
```
$(call GPU_ARCH_CODE,compute_30,compute_30)
```
with
```
-generate-code arch=compute_30,code=compute_30 
```
That way the compiler will issue an error when that combination is no longer supported.
Thoughts @pshipton?

Reply:
Can you check if GPU_ARCH_CODES is empty?

Reply:
With no `--generate-code` options, nvcc seems to behave reasonably. CUDA 9 works as if the following were specified.
```
--generate-code arch=compute_30,code=compute_30
--generate-code arch=compute_30,code=sm_30
```
This yields code that will be applicable to all devices with compute capability 3 or higher.

Reply:
Wouldn't it be better if the 35 kernel were generated?
If this code ever breaks, how will it be detected?

Reply:
Better only on machines with a GPU whose compute capability is exactly 3.5.
The PTX code requested via `...,code=compute_30` would be used for all other GPUs with compute capability >= 3.0.

Reply:
If this were to break, com.ibm.gpu.Maths.sortArray() will throw exceptions when using an unsupported GPU.

Reply:
Given that gpu machines with all the different compute capabilities may not be available for testing, wouldn't it be better to discover there is a problem at build time than to wait for a customer to find an issue or discover that performance isn't as good as it should be?

Reply:
Better than verifying that GPU_ARCH_CODES is not empty, I will add a check that at least one of compute_20 or compute_30 is in GPU_ARCH_LIST which will imply that we will have at least one of the following options:
```
--generate-code arch=compute_20,code=compute_20
--generate-code arch=compute_30,code=compute_30
```
That will ensure that all devices with compute capability at least 2.0 (or 3.0) will be usable.